cmd/device-definitions-api/fuel_image_service.go

```diff
@@ -0,0 +1,195 @@
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	return nil
+}
+
+func NewFuelService(ctx context.Context, settings *config.Settings, logger *zerolog.Logger) *FuelServiceAPI {
+
+	pdb := db.NewDbConnectionFromSettings(ctx, &settings.DB, true)
+	pdb.WaitForDB(*logger)
+
+	return &FuelServiceAPI{
+		VehicleURL: settings.FuelAPIVehiclesEndpoint,
+		ImageURL:   settings.FuelAPIImagesEndpoint,
+		Key:        settings.FuelAPIKey,
+		log:        logger,
+		db:         pdb.DBS(),
+	}
+}
+
+func (fs *FuelServiceAPI) writeToTable(ctx context.Context, data []deviceData, prodID int, prodFormat int) error {
+
+	for _, d := range data {
+		for n := range d.Models {
+
+			img, err := fs.fetchDeviceImage(d.Make, d.Models[n].Model, d.Models[n].Year, prodID, prodFormat)
+			if err != nil {
+				continue
+			}
+
+			var p models.Image
+			p.ID = ksuid.New().String()
+			p.DeviceDefinitionID = d.Models[n].DeviceDefinitionID
+			p.FuelAPIID = null.StringFrom(img.FuelAPIID)
+			p.Width = null.IntFrom(img.Width)
+			p.Height = null.IntFrom(img.Height)
+			p.SourceURL = img.SourceURL
+			p.DimoS3URL = null.StringFrom("")
+			p.Color = img.Color
+
+			err = p.Upsert(ctx, fs.db.Writer, false, []string{models.ImageColumns.DeviceDefinitionID, models.ImageColumns.SourceURL}, boil.Infer(), boil.Infer())
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (fs *FuelServiceAPI) fetchDeviceImage(mk, mdl string, yr int, prodID int, prodFormat int) (image, error) {
+
+	img, err := fs.imageRequest(mk, mdl, yr, prodID, prodFormat)
+	if err != nil {
+		return image{}, err
+	}
+
+	if img.SourceURL == "" {
+		img, err = fs.imageRequest(mk, mdl, 0, prodID, prodFormat)
+		if err != nil {
+			return image{}, err
+		}
+
+		if img.SourceURL == "" {
+			m := strings.Split(mdl, " ")
+			img, err = fs.imageRequest(mk, m[0], 0, prodID, prodFormat)
+			if err != nil {
+				return image{}, err
+			}
+		}
+	}
+
+	if img.SourceURL == "" {
+		fs.log.Log().Msgf("request for device image unsuccessful: %s %s %d", mk, mdl, yr)
+		return image{}, errors.New("request for device image unsuccessful")
+	}
+
+	return img, nil
+
+}
+
+func (fs *FuelServiceAPI) imageRequest(mk, mdl string, yr int, prodID int, prodFormat int) (image, error) {
+	vehicleReqURL := fmt.Sprintf("?year=%d&make=%s&model=%s&api_key=%s", yr, mk, mdl, fs.Key)
+	vehicleResp, err := http.Get(fs.VehicleURL + vehicleReqURL)
+	if err != nil {
+		return image{}, err
+	}
+	vehicleData, err := ioutil.ReadAll(vehicleResp.Body)
+	if err != nil {
+		return image{}, err
+	}
+	vehicleID := gjson.Get(string(vehicleData), "0.id").Str
+	imageReqURL := fmt.Sprintf("/%s?api_key=%s&productID=%d&productFormatIDs=%d", vehicleID, fs.Key, prodID, prodFormat)
+	imageResp, err := http.Get(fs.ImageURL + imageReqURL)
+	if err != nil {
+		return image{}, err
+	}
+	imageData, err := ioutil.ReadAll(imageResp.Body)
+	if err != nil {
+		return image{}, err
+	}
+	imageURL := gjson.Get(string(imageData), "products.0.productFormats.0.assets.0.url").Str
```

Comment:
- if there is more than one image, eg. i think it returned multiple colors, should we return an array of images? I understand we may not need it all right now, so if too much to handle just leave a TODO comment
- what happens if there are no images returned in the json body? or would we get a different http status code when that happens?

Reply:
updated to grab array of all image urls returned 
https://github.com/DIMO-Network/device-definitions-api/blob/e8a53e6cd46a09a01963906123c180c27b409162/cmd/device-definitions-api/fuel_image_service.go#L192-L198

Reply:
if no images are returned, we'll get a 404 error